engine/classes/Elgg/I18n/Translator.php

```diff
@@ -402,33 +403,52 @@ public function registerTranslations($path, $load_all = false, $language = null)
 			$load_language_files = array_unique($load_language_files);
 		}
 
-		$handle = opendir($path);
-		if (!$handle) {
-			_elgg_services()->logger->error("Could not open language path: $path");
-			return false;
-		}
-
 		$return = true;
-		while (false !== ($language_file = readdir($handle))) {
-			// ignore bad files
-			if (substr($language_file, 0, 1) == '.' || substr($language_file, -4) !== '.php') {
-				continue;
+		if (!$load_all) {
+			foreach ($load_language_files as $language_file) {
+				$return = $return && $this->includeLanguageFile($path . $language_file);
 			}
-
-			if (in_array($language_file, $load_language_files) || $load_all) {
-				$result = (include $path . $language_file);
-				if ($result === false) {
-					$return = false;
+		} else if ($handle = opendir($path)) {
+			while (false !== ($language_file = readdir($handle))) {
+				// ignore bad files
+				if (substr($language_file, 0, 1) == '.' || substr($language_file, -4) !== '.php') {
 					continue;
-				} elseif (is_array($result)) {
-					$this->addTranslation(basename($language_file, '.php'), $result);
+				}
+
+				if (in_array($language_file, $load_language_files) || $load_all) {
+					$return = $return && $this->includeLanguageFile($path . $language_file);
 				}
 			}
+			closedir($handle);
+		} else {
+			_elgg_services()->logger->error("Could not open language path: $path");
+			$return = false;
 		}
 
 		return $return;
 	}
 
+	/**
+	 * Load cached or include a language file by its path
+	 *
+	 * @param string $path Path to file
+	 * @return bool
+	 */
+	protected function includeLanguageFile($path) {
+		$cache_key = "lang/" . sha1($path);
+		$result = elgg_get_system_cache()->load($cache_key);
```

Comment:
Isn't this more costly than the file read?

Reply:
Not if there is persistent cache. When it's a file system cache, I think it's the same thing.

Reply:
On an average page load this won't have much effect, on tests however it gives a decent boost.

Reply:
I see no effect on test runtime...

Reply:
Try it locally

Reply:
And travis is no indicator, build times have been incredibly volatile

Reply:
Also this is what fixed the phpdbg memory issues.